File: DataStuctures/Stack/Implementation/stack.py
class Node:
    def __init__(self, data, next=None):
        self.data = data
        self.next = next


class Stack:
    def __init__(self):
        self.top = None

    def pop(self):
        """
        Remove first item from stack
        """
        if self.top is None:
            raise IndexError("Stack is empty")
        node = self.top
        data = node.data
        self.top = node.next
        del node
        return data

    def peek(self):
        """
        Return first stack node data
        """
        if self.top is None:
            return None
        return self.top.data

    def push(self, data):
        """
        Create a new node and fill it with given data, and then push it into stack.
        """
        node = Node(data)
        node.next = self.top
        self.top = node

    def __len__(self):
        """
        Return Length of the stack
        """
        if self.top is None:
            return 0
        temp = self.top
        i = 1
        while temp.next is not None:
            i += 1
            temp = temp.next
        return i
    
    def __str__(self):
        """
        Return str object of stack nodes data.
        If stack is empty return str obj equal to "None".
        Example:
        # If We have a stack that has 3 items below, the output will be:
        1 -> 2 -> 3 -> None
        """
        output = ""
        temp = self.top
        while temp is not None:
            output += f"{temp.data} -> "
            temp = temp.next
        output += "None"
        return output
    
    def clean(self):
        """
        Empty the stack whenever you call it.
        """
        while self.top is not None:
            t = self.top
            self.top = t.next
            del t
        

File: DataStuctures/Stack/Implementation/stack.cpp
#include <iostream>
using namespace std;

struct Node
{
    string data;
    Node *next;
};

void push(Node*& top, string data){
    // Create & Append nodes into stack with given data

    Node* new_node = new Node{data, nullptr};
    new_node->next = top;
    top = new_node;
}

string peek(Node*& top){
    // Return top node value

    if (top != nullptr){
        return top->data;
    }
    return "null";
}

string pop(Node*& top){
    /*
    This function will remove top node and return the data.
    If stack was empty, return null.
    */
    if (top != nullptr){
        string data = top->data;
        Node* temp = top;
        top = top->next;
        delete temp;
        return data;
    }
    return "null";
}

void clear(Node*& top){
    /*
    This method will remove all nodes from stack and make it empty.
    */
    while (top != nullptr){
        Node* temp = top;
        top = top->next;
        delete temp;
    }
}

void display(Node*& top){
    /* 
    Print stack node data in console
    If stack was empty just cout null 
    */

    Node* node = top;
    while (node != nullptr){
        cout << node->data << " -> ";    
        node = node->next;
    }
    cout << "null" << endl;
}

int main(){
    // Put Codes in here
    // ...

    return 0;
}